### ISSUE TITLE
Low performance for encoding/decoding when using custom member names transformation with semiauto

### ISSUE THREAD

**Issue by vndzzz** [circe/circe#849]
The issue is occured when member names should be rewritten, i.e. with using `io.circe.generic.extras.Configuration.default.withSnakeCaseMemberNames` regexp transformation are occured on every json encoding/decoding (see `io.circe.generic.extras.ConfiguredDecoder`, `io.circe.generic.extras.ConfiguredObjectEncoder`). 
I think they should be performed only once when encoder/decoder are derived.

**Comment by alex-vv** [comment]
+1

I had the same issue. As a workaround I had to cache these member names:

```
private val fieldNamesCache = TrieMap[String, String]()

implicit val config: Configuration = Configuration.default.copy(transformMemberNames = str =>
  fieldNamesCache.getOrElseUpdate(str, Configuration.snakeCaseTransformation(str)))
```

**Comment by travisbrown** [comment]
Good catch, thanks. I'll happily accept a fix, otherwise will try to get it in this weekend.

**Comment by vndzzz** [comment]
Please review my first PR #860 It should solve this issue.

**Comment by travisbrown** [comment]
Resolved by #860.
